Fold header lines with the correct length

danielabyan opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | yes
| New Feature   | no
| RFC           | yes
| QA            | no

### Description

According to the RFC 2822 specification (clause 2.1.1), the header length should be no more than 78 characters.

> Each line of characters MUST be no more than 998 characters, and SHOULD be no more than 78 characters, excluding the CRLF.

According to the specification, the length of the ENTIRE LINE should not exceed 78 characters. Currently, the `laminas-mime` library only counts the length of a **value** of the header. 

### Reproducing the problem
```
$message = new Message();
$message->setEncoding('UTF-8');
$message->addFrom('matthew@example.org', 'Matthew Somelli');
$message->addTo('foobar@example.com');
$message->setSubject('xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx');
$message->setBody('This is the message body.');

echo $message->toString();
```

This code will return the following result.

> Date: Thu, 21 Oct 2021 10:37:01 +0000
> From: =?UTF-8?Q?Matthew=20Somelli?= <matthew@example.org>
> To: foobar@example.com
> Subject: =?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20?=
>  =?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx?=
> 
> This is the message body.

As you can see from the example, the header is longer than 78 characters.

### New code behavior

This commit has fixed this issue. In this pool request, the code above will return the following result.

> Date: Thu, 21 Oct 2021 10:38:56 +0000
> From: =?UTF-8?Q?Matthew=20Somelli?= <matthew@example.org>
> To: foobar@example.com
> Subject: =?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20?=
>  =?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx?=
> 
> This is the message body.

Now, the title does not exceed 78 characters.

### Note
I will wait until this pool request is merged to open my pool request on [laminas-mail](https://github.com/laminas/laminas-mail/pull/174)

danielabyan committed:
Fold a header line with the correct length

According to the RFC 2822 specification (clause 2.1.1),
the header length should be no more than 78 characters.
   "Each line of characters MUST be no more than 998
   characters, and SHOULD be no more than 78 characters,
   excluding the CRLF."
According to the specification, the length of the ENTIRE
LINE should not exceed 78 characters. Currently, the
`laminas-mail` library only counts the length of a value
of the header. This commit has fixed this issue.

Signed-off-by: danielabyan <danielabyan@gmail.com>

Ocramius reviewed (changes requested):
This kind of change requires careful test additions.

glensc commented:
Perhaps it's just better to add a new api that can take headerline or headerkey,headervalue pair as input?

or maybe that is not needed if the caller just passes headerline instead of headervalue as input?
it may break headerkeys that are longer than 998 bytes, but perhaps caller should handle (throw) such edge cases themselves?

danielabyan commented:
@glensc I was considering passing the whole string, but there is a problem here. If we pass the entire line, the header name will be encoded as well. But according to the RFC 2822, only the value (body) of the header must be encoded. 
For example, this code
```
$mime = Mime\Mime::encodeQuotedPrintableHeader(
    'Subject: xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx xxxxx',
    'UTF-8',
    78
);

var_dump($mime);
```

Will return a result like this
```
=?UTF-8?Q?Subject:=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20?=
 =?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx?="
```

As you can see, in this case, the whole line breaks. So, I made the changes to be minimal and not to break past behavior.

danielabyan commented:
@Ocramius Do we have any updates? I just need this change to improve my mailing since my emails have long lines than recommended in RFC.

Ocramius commented on test/MimeTest.php:
```diff
@@ -152,6 +152,53 @@ public static function dataTestEncodeMailHeaderQuotedPrintable(): array
+                "=?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20?=" . Mime\Mime::LINEEND
+                . " =?UTF-8?Q?xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx=20xxxxx?=",
+                9,
+            ], // Long string with header name size
```
Not a problem, but if you use these as array keys, instead of comments, you get valuable PHPUnit output while running (`--testdox` or such) and in case of failures :)

danielabyan replied:
Fixed ))

Ocramius commented on src/Mime.php:
```diff
@@ -493,15 +493,21 @@ private static function _encodeQuotedPrintable($str)
         $lineLength = self::LINELENGTH,
-        $lineEnd = self::LINEEND
+        $lineEnd = self::LINEEND,
+        $headerNameSize = 0
```
Are there any consumers in `laminas/*` that need to be updated to handle this?

danielabyan replied:
Yes, we use this method in the lamins-mail. I have already created a pool request [there](https://github.com/laminas/laminas-mail/pull/174). When this request is merged, I will send it for review.

Ocramius commented on src/Mime.php:
```diff
@@ -493,15 +493,21 @@ private static function _encodeQuotedPrintable($str)
-     * @param string $lineEnd Defaults to {@link LINEEND}
+     * @param int    $lineLength       Defaults to {@link LINELENGTH}
+     * @param string $lineEnd          Defaults to {@link LINEEND}
+     * @param int    $headerNameSize   When folding a line, it is necessary to calculate
```
Potentially worth documenting `positive-int|0` here

danielabyan replied:
Fixed

Ocramius reviewed (approved):
LGTM :+1: 

@glensc can I have a second opinion? Code-wise, it makes sense, but I don't know much about RFC 2822

danielabyan commented:
@Ocramius Sorry for writing a lot about this, but can we release it if everything is fine with this request?

Ocramius commented:
Given my low RFC-fu here, perhaps either @Xerkus or @glensc can give a green light here.

Code-wise, it is fine for me.

Xerkus commented:
Makes sense to me. Spec does speak about line length irrespective to the structured content of the line.

danielabyan commented:
I have made all the changes in the code to meet the requirements of the code sniffer. But now I have a problem with composer. Please check [this problem](https://github.com/laminas/laminas-mime/runs/7859720530?check_suite_focus=true).  
![image](https://user-images.githubusercontent.com/25223574/184900446-fde16a9d-c2d5-442a-b785-02bbe520ff6c.png)
As you can see the problem lies in the new security policy in [Composer 2.2](https://getcomposer.org/doc/06-config.md#allow-plugins).
This problem can be fixed if I add this configuration to the "composer.json" file
```
   ...
    "config": {
        "sort-packages": true,
        "allow-plugins": {
            "dealerdirect/phpcodesniffer-composer-installer": true
        }
    },
   ...
```
@Ocramius  can I do this in this pull request or should I open a separate pull request? 

Thank you.

Ocramius commented:
@danielabyan from my point of view, you can do it here :+1:

danielabyan committed:
Introduce `allow-plugins` config in the composer.json

Changes have been made to meet the requirements of
composer 2.2. Composer 2.2 introduced a new security
layer to prevent executing arbitrary code when the
Composer runs. We have a library
`dealerdirect/phpcodesniffer-composer-installer` which
is executed during the installation of the project.
In this commit, this library was allowed to run during
the installation of the project.

Signed-off-by: danielabyan <danielabyan@gmail.com>

danielabyan committed:
Update phpunit up to `9.5` version

In the old version of PHPUnit there is a bug that
does not allow you to build the project. An error
`Schemas parser error : Failed to locate the main schema
resource at 'vendor/phpunit/phpunit/phpunit.xsd'` is
thrown during the test execution. This bug has been
fixed in version `9.5`. Therefore, the version of the
library has been updated.

Signed-off-by: danielabyan <danielabyan@gmail.com>

Ocramius commented on composer.json:
```diff
@@ -28,7 +28,7 @@
         "laminas/laminas-coding-standard": "~2.2.1",
         "laminas/laminas-mail": "^2.12",
-        "phpunit/phpunit": "^9.3"
+        "phpunit/phpunit": "9.5"
```
This should be `^9.5`

danielabyan replied:
Will fix it now

danielabyan replied:
Done

danielabyan commented:
@Ocramius I fixed all the bugs and issues that were in this pool request. Please check my work, and if everything is correct, please create a new release, as some mail services refuse to deliver my emails due to RFC complaints. Thanks a lot.

Ocramius reviewed (approved):
Thanks @danielabyan!
